sync_check/diff.go

```diff
@@ -0,0 +1,677 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"database/sql"
+	"fmt"
+	"math/rand"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/ngaut/log"
+	"github.com/pingcap/tidb-tools/pkg/db"
+	"github.com/pingcap/tidb-tools/sync_check/util"
+	"github.com/pingcap/tidb/model"
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/types"
+)
+
+// Diff contains two sql DB, used for comparing.
+type Diff struct {
+	db1          *sql.DB
+	db2          *sql.DB
+	dbName       string
+	chunkSize    int
+	sample       int
+	checkThCount int
+	useRowID     bool
+	tables       []*TableCheckCfg
+	fixSqlFile   *os.File
+	sqlCh        chan string
+	wg           sync.WaitGroup
+}
+
+// NewDiff returns a Diff instance.
+func NewDiff(db1, db2 *sql.DB, dbName string, chunkSize, sample, checkThCount int,
+	useRowID bool, tables []*TableCheckCfg, filename, snapshot string) (diff *Diff, err error) {
+	diff = &Diff{
+		db1:          db1,
+		db2:          db2,
+		dbName:       dbName,
+		chunkSize:    chunkSize,
+		sample:       sample,
+		checkThCount: checkThCount,
+		useRowID:     useRowID,
+		tables:       tables,
+		sqlCh:        make(chan string),
+	}
+	for _, table := range diff.tables {
+		table.Info, err = pkgdb.GetSchemaTableWithRowID(diff.db1, diff.dbName, table.Name, useRowID)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		table.Schema = diff.dbName
+	}
+	diff.fixSqlFile, err = os.Create(filename)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	if snapshot != "" {
+		err = pkgdb.SetSnapshot(db1, snapshot)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+
+	return diff, nil
+}
+
+// Equal tests whether two database have same data and schema.
+func (df *Diff) Equal() (equal bool, err error) {
+	defer func() {
+		df.fixSqlFile.Close()
+		df.db1.Close()
+		df.db2.Close()
+	}()
+
+	df.wg.Add(1)
+	go df.WriteSqls()
+
+	equal = true
+	tbls1, err := pkgdb.GetTables(df.db1, df.dbName)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+
+	tbls2, err := pkgdb.GetTables(df.db2, df.dbName)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+
+	eq := equalStrings(tbls1, tbls2)
+	// len(df.tables) == 0 means check all tables
+	if !eq && len(df.tables) == 0 {
+		log.Errorf("show tables get different table. [source db tables] %v [target db tables] %v", tbls1, tbls2)
+		equal = false
+	}
+
+	if len(df.tables) == 0 {
+		df.tables = make([]*TableCheckCfg, 0, len(tbls1))
+		for _, name := range tbls1 {
+			table := &TableCheckCfg{Name: name, Schema: df.dbName}
+			table.Info, err = pkgdb.GetSchemaTableWithRowID(df.db1, df.dbName, name, df.useRowID)
+			if err != nil {
+				return false, errors.Trace(err)
+			}
+
+			df.tables = append(df.tables, table)
+		}
+	}
+
+	for _, table := range df.tables {
+		eq, err = df.EqualIndex(table.Name)
```

Comment:
do it parallelly?

Reply:
I split a table data check job to multiple jobs, and do these jobs in multiple groutine. I think it's enough, or I can refine this code next pr.